pkg/controller/storagecluster/placement_test.go

```diff
@@ -13,108 +13,105 @@ import (
+	cases := []struct {
+		placements        rookv1.PlacementSpec
+		labelSelector     *metav1.LabelSelector
+		expectedPlacement rookv1.Placement
+	}{
+		{
+			// Case 1: Defaults are preserved i.e no placement and no label selector
+			placements:    rookv1.PlacementSpec{},
+			labelSelector: nil,
+			expectedPlacement: rookv1.Placement{
+				NodeAffinity: defaults.DefaultNodeAffinity,
+				Tolerations:  defaults.DaemonPlacements["all"].Tolerations,
 			},
 		},
-	}
-	assert.Equal(t, mockPlacements["all"], getPlacement(sc, "all"))
-
-	// Case 4: The LabelSelector modifies the configured Placements correctly
-	sc = &ocsv1.StorageCluster{}
-	mockStorageCluster.DeepCopyInto(sc)
-	sc.Spec.Placement = mockPlacements
-	sc.Spec.LabelSelector = &metav1.LabelSelector{
-		MatchExpressions: []metav1.LabelSelectorRequirement{
-			metav1.LabelSelectorRequirement{
-				Key:      MasterAffinityKey,
-				Operator: metav1.LabelSelectorOpExists,
+		{
+			// Case 2: The configured Placements override the defaults
+			placements:    emptyPlacements,
+			labelSelector: nil,
+			expectedPlacement: rookv1.Placement{
+				NodeAffinity: defaults.DefaultNodeAffinity,
+			},
+		},
+		{
+			// Case 3: LabelSelector to modify the default Placements correctly
+			placements:    rookv1.PlacementSpec{},
+			labelSelector: &workerLabelSelector,
+			expectedPlacement: rookv1.Placement{
+				NodeAffinity: &corev1.NodeAffinity{
+					RequiredDuringSchedulingIgnoredDuringExecution: &workerNodeSelector,
+				},
+				Tolerations: defaults.DaemonPlacements["all"].Tolerations,
+			},
+		},
+		{
+			// Case 4: The LabelSelector modifies the configured Placements correctly
+			placements:    emptyPlacements,
+			labelSelector: &workerLabelSelector,
+			expectedPlacement: rookv1.Placement{
+				NodeAffinity: &corev1.NodeAffinity{
+					RequiredDuringSchedulingIgnoredDuringExecution: &workerNodeSelector,
+				},
+			},
+		},
+		{
+			// Case 5: Empty LabelSelector sets no NodeAffinity
+			placements:    rookv1.PlacementSpec{},
+			labelSelector: &emptyLabelSelector,
+			expectedPlacement: rookv1.Placement{
+				Tolerations: defaults.DaemonPlacements["all"].Tolerations,
 			},
 		},
 	}
-	expectedPlacements := mockPlacements
-	expectedPlacements["all"].NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms[0].MatchExpressions[0].Key = MasterAffinityKey
-	assert.Equal(t, expectedPlacements["all"], getPlacement(sc, "all"))
 
-	// Case 5: Empty LabelSelector sets no Node Affinity
-	sc = &ocsv1.StorageCluster{}
-	mockStorageCluster.DeepCopyInto(sc)
-	sc.Spec.LabelSelector = &metav1.LabelSelector{
-		MatchExpressions: []metav1.LabelSelectorRequirement{},
+	for _, c := range cases {
+		sc := &ocsv1.StorageCluster{}
+		mockStorageCluster.DeepCopyInto(sc)
+		sc.Spec.Placement = c.placements
+		sc.Spec.LabelSelector = c.labelSelector
+		expectedPlacement := c.expectedPlacement
+		assert.Equal(t, expectedPlacement, getPlacement(sc, "all"))
```

Comment:
Better split the invocation of `getPlacement` into a separate line and assign the result to a variable. It will then be easier to see what happens if getPlacement crashes, etc.

Reply:
Sure.